docs/advanced/pycpp/object.rst

```diff
@@ -62,6 +64,40 @@ This example obtains a reference to the Python ``Decimal`` class.
     py::object scipy = py::module::import("scipy");
     return scipy.attr("__version__");
 
+.. _implicit_casting:
+
+Implicit (automatic) casting
+============================
+
+Instead of using a generic :class:`object` as return type, it is possible to
```

Comment:
This start of the paragraph is missing the answer to "When?".
What about e.g., "When using this C++ interface for Python types or calling Python functions, and `py::object` are returned, [...]. The same holds for the proxy objects returned from `operator[]` or `attr`."?

This might actually be a reason to move this down a few more sections, such that it's after the "Calling Python functions" and "Calling Python methods" (and then maybe after the arguments sections, because they're related?). But then again, there will always be this circular dependency, so you could just put a forward reference. I'll leave this up to your judgement.

Reply:
Hm you're right, it fits better a bit farther below. Also because the subsequent sections use `py::object` throughout. I've moved it down.

Reply:
And thanks for the text suggestion, I've used parts of it.